File: problems/c7-shannon/compute/search_cosets.py
#!/usr/bin/env python3
"""Pack cosets of a good 2-dimensional F_7-code.

If V <= (Z/7)^5 is 2-dimensional and V ∩ {-1,0,1}^5 = {0}, then a union of
cosets a+V is independent iff the quotient points are independent in the
Cayley graph of F_7^5/V with connection set ( {-1,0,1}^5 + V )/V.
That graph has 343 vertices. Eight independent cosets would give 392 points.
"""

from __future__ import annotations

import itertools
import random
import time
from pathlib import Path

from c7_common import DIM, N, encode, format_word, residual_of
from verify_set import first_conflict

HERE = Path(__file__).resolve().parent
SMALL = {0, 1, 6}


def apply(rows, coeffs):
    out = [0] * DIM
    for c, row in zip(coeffs, rows):
        if not c:
            continue
        for j in range(DIM):
            out[j] = (out[j] + c * row[j]) % N
    return tuple(out)


def is_good_2(rows) -> bool:
    for coeffs in itertools.product(range(N), repeat=2):
        if coeffs == (0, 0):
            continue
        w = apply(rows, coeffs)
        if all(x in SMALL for x in w):
            return False
    return True


def span2(rows):
    return [apply(rows, (a, b)) for a in range(N) for b in range(N)]


def complete_basis(rows):
    """Extend 2 rows to an F7-basis of F7^5. Return 3 complementary rows."""
    mat = [list(r) for r in rows]
    used = [False] * 5
    # mark pivot-ish columns
    r = 0
    tmp = [list(x) for x in mat]
    for c in range(5):
        piv = None
        for i in range(r, 2):
            if tmp[i][c] % N:
                piv = i
                break
        if piv is None:
            continue
        tmp[r], tmp[piv] = tmp[piv], tmp[r]
        inv = pow(tmp[r][c], -1, N)
        tmp[r] = [(x * inv) % N for x in tmp[r]]
        for i in range(2):
            if i == r:
                continue
            f = tmp[i][c]
            if f:
                tmp[i] = [(tmp[i][j] - f * tmp[r][j]) % N for j in range(5)]
        used[c] = True
        r += 1
        if r == 2:
            break
    comps = []
    for c in range(5):
        if used[c]:
            continue
        e = [0] * 5
        e[c] = 1
        comps.append(tuple(e))
    # may have 3 standard basis vectors not used; if fewer, fill remaining
    while len(comps) < 3:
        for c in range(5):
            e = [0] * 5
            e[c] = 1
            if tuple(e) not in comps and tuple(e) not in rows:
                comps.append(tuple(e))
                break
    return comps[:3]


def quotient_index(v, comps):
    """Coordinates of v along the complementary 3-space, assuming V-coords ignored.

    We identify F7^5 / V with F7^3 via a complementary subspace W. This is only
    a bijection if V ⊕ W = whole space. We check that by rank.
    """
    # solve v = a0 r0 + a1 r1 + b0 c0 + b1 c1 + b2 c2
    # We only need (b0,b1,b2) as the quotient coordinate if the 5x5 is invertible.
    return None  # filled in after we build the 5x5


def invert5(rows5):
    # Gaussian invert over F7. rows5 is 5 tuples. Return inverse as list of lists.
    a = [list(r) + [1 if i == j else 0 for j in range(5)] for i, r in enumerate(rows5)]
    for col in range(5):
        piv = None
        for i in range(col, 5):
            if a[i][col] % N:
                piv = i
                break
        if piv is None:
            return None
        a[col], a[piv] = a[piv], a[col]
        inv = pow(a[col][col], -1, N)
        a[col] = [(x * inv) % N for x in a[col]]
        for i in range(5):
            if i == col:
                continue
            f = a[i][col]
            if f:
                a[i] = [(a[i][j] - f * a[col][j]) % N for j in range(10)]
    return [row[5:] for row in a]


def matvec(M, v):
    return tuple(sum(M[i][j] * v[j] for j in range(5)) % N for i in range(len(M)))


def greedy_mis_graph(n: int, adj: list[int], rng: random.Random) -> list[int]:
    order = list(range(n))
    rng.shuffle(order)
    taken = []
    banned = 0
    # adj[i] is a bitset of neighbors (n=343 fits in a Python int)
    for v in order:
        if (banned >> v) & 1:
            continue
        taken.append(v)
        banned |= adj[v] | (1 << v)
    return taken


def local_improve(n: int, adj: list[int], taken: list[int], rounds: int, rng: random.Random) -> list[int]:
    s = set(taken)
    for _ in range(rounds):
        # 1-out: remove one, add as many as possible
        if not s:
            break
        v = rng.choice(list(s))
        s.remove(v)
        banned = 0
        for u in s:
            banned |= adj[u] | (1 << u)
        extras = [i for i in range(n) if not ((banned >> i) & 1)]
        rng.shuffle(extras)
        added = []
        for i in extras:
            if (banned >> i) & 1:
                continue
            added.append(i)
            banned |= adj[i] | (1 << i)
        if len(s) + len(added) >= len(taken):
            s.update(added)
            taken = list(s)
        else:
            s.add(v)
    return taken


def main() -> None:
    t0 = time.time()
    rng = random.Random(0)
    # Enumerate RREF 2x5
    mats = []
    for pivots in itertools.combinations(range(5), 2):
        free = [j for j in range(5) if j not in pivots]
        for fill in itertools.product(range(N), repeat=2 * len(free)):
            rows = []
            idx = 0
            for p in pivots:
                row = [0] * 5
                row[p] = 1
                for j in free:
                    row[j] = fill[idx]
                    idx += 1
                rows.append(tuple(row))
            mats.append(tuple(rows))
    print(f"2x5 rref {len(mats)}", flush=True)
    best_cosets = 0
    best_total = 0
    n_good = 0
    lines = []
    checked = 0
    for rows in mats:
        checked += 1
        if not is_good_2(rows):
            if checked % 100000 == 0:
                print(f"checked {checked} good={n_good} best_cosets={best_cosets}", flush=True)
            continue
        n_good += 1
        comps = []
        # complementary standard basis columns
        piv = [i for i, x in enumerate(rows[0]) if x == 1][:1]
        # simpler: try standard basis triples until 5x5 invertible
        inv = None
        chosen = None
        for triple in itertools.combinations(range(5), 3):
            cand = [tuple(1 if j == t else 0 for j in range(5)) for t in triple]
            M = list(rows) + cand
            invM = invert5(M)
            if invM is not None:
                chosen = cand
                inv = invM
                break
        if inv is None:
            continue
        # quotient coord of a vector x is the last 3 coords of inv * x
        def qcoord(x):
            y = matvec(inv, x)
            return y[2] * 49 + y[3] * 7 + y[4]

        # forbidden quotient points: images of {-1,0,1}^5
        forbidden = set()
        for deltas in itertools.product((-1, 0, 1), repeat=5):
            forbidden.add(qcoord(tuple(d % N for d in deltas)))
        # Cayley graph on 343 points, edge if difference in forbidden\{0}
        adj = [0] * 343
        forb = [f for f in forbidden if f != 0]
        for i in range(343):
            bits = 0
            for f in forb:
                # difference in F7^3
                a0, r = divmod(i, 49)
                a1, a2 = divmod(r, 7)
                b0, r = divmod(f, 49)
                b1, b2 = divmod(r, 7)
                d0 = (a0 - b0) % 7
                d1 = (a1 - b1) % 7
                d2 = (a2 - b2) % 7
                j = d0 * 49 + d1 * 7 + d2
                if j != i:
                    bits |= 1 << j
            adj[i] = bits
        # several greedy + local
        best_here = []
        for trial in range(12):
            pack = greedy_mis_graph(343, adj, rng)
            pack = local_improve(343, adj, pack, 80, rng)
            if len(pack) > len(best_here):
                best_here = pack
        ncos = len(best_here)
        total = 49 * ncos
        if ncos > best_cosets or total > best_total:
            best_cosets = max(best_cosets, ncos)
            best_total = max(best_total, total)
            print(f"good V={rows} forbidden={len(forbidden)} cosets={ncos} total={total}", flush=True)
            lines.append(f"{rows} forb={len(forbidden)} cosets={ncos} total={total}")
        if total >= 368:
            # materialize the union
            V = span2(rows)
            # map quotient index -> a representative in W
            reps = {}
            for b in itertools.product(range(N), repeat=3):
                w = apply(chosen, b)
                reps[qcoord(w)] = w
            pts = []
            for qi in best_here:
                w = reps[qi]
                for v in V:
                    pts.append(encode(tuple((w[j] + v[j]) % N for j in range(5))))
            pts = sorted(set(pts))
            if first_conflict(pts) is None:
                out = HERE / f"R{len(pts)}_cosets.txt"
                out.write_text("\n".join(format_word(v) for v in pts) + "\n")
                print(f"WROTE {out} size={len(pts)}")
                (HERE / "coset_search.txt").write_text(
                    f"HIT size {len(pts)}\n" + "\n".join(lines) + "\n"
                )
                return
        if n_good <= 3 or n_good % 20 == 0:
            print(f"  good={n_good} checked={checked} best_cosets={best_cosets}", flush=True)
    (HERE / "coset_search.txt").write_text(
        f"good {n_good}\nbest_cosets {best_cosets}\nbest_total {best_total}\n"
        f"seconds {time.time()-t0:.1f}\n" + "\n".join(lines) + "\n"
    )
    print(f"done good={n_good} best_cosets={best_cosets} best_total={best_total} in {time.time()-t0:.1f}s")


if __name__ == "__main__":
    main()

File: problems/c7-shannon/compute/q1/search_cosets.c
/* Pack cosets of a good 2-dimensional F7-subspace (shape 4).

   V ≤ F7^5, dim 2, V ∩ {-1,0,1}^5 = {0}. The quotient F7^5 / V has 343
   points. An independent set of 8 quotient points is 392 vertices.

   gcc -O3 -o q1/search_cosets q1/search_cosets.c
   ./q1/search_cosets
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define N 7
#define DIM 5
#define NV 16807
#define QN 343

static int SMALL[243][DIM];
static int nsmall;

static int rank2_ok(const int a[DIM], const int b[DIM]) {
    /* not scalar multiples */
    int seen = 0, ratio = -1;
    for (int i = 0; i < DIM; i++) {
        if (a[i] == 0 && b[i] == 0) continue;
        if (a[i] == 0 || b[i] == 0) return 1;
        int r = 0;
        /* b[i] = r * a[i] */
        for (int t = 0; t < 7; t++)
            if ((t * a[i]) % 7 == b[i]) {
                r = t;
                break;
            }
        if (!seen) {
            ratio = r;
            seen = 1;
        } else if (r != ratio)
            return 1;
    }
    return 0;
}

static int is_good(const int a[DIM], const int b[DIM]) {
    for (int s = 0; s < 7; s++)
        for (int t = 0; t < 7; t++) {
            if (s == 0 && t == 0) continue;
            int ok_small = 1;
            for (int i = 0; i < DIM; i++) {
                int x = (s * a[i] + t * b[i]) % 7;
                if (x > 1 && x < 6) {
                    ok_small = 0;
                    break;
                }
            }
            if (ok_small) return 0;
        }
    return 1;
}

static int invert5(int M[5][5], int Inv[5][5]) {
    int a[5][10];
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) a[i][j] = M[i][j];
        for (int j = 0; j < 5; j++) a[i][5 + j] = (i == j);
    }
    for (int col = 0; col < 5; col++) {
        int piv = -1;
        for (int i = col; i < 5; i++)
            if (a[i][col] % 7) {
                piv = i;
                break;
            }
        if (piv < 0) return 0;
        if (piv != col) {
            for (int j = 0; j < 10; j++) {
                int tmp = a[col][j];
                a[col][j] = a[piv][j];
                a[piv][j] = tmp;
            }
        }
        int inv = 0;
        for (int t = 1; t < 7; t++)
            if ((t * a[col][col]) % 7 == 1) {
                inv = t;
                break;
            }
        for (int j = 0; j < 10; j++) a[col][j] = (a[col][j] * inv) % 7;
        for (int i = 0; i < 5; i++) {
            if (i == col) continue;
            int f = a[i][col];
            if (!f) continue;
            for (int j = 0; j < 10; j++)
                a[i][j] = (a[i][j] - f * a[col][j] % 7 + 14) % 7;
        }
    }
    for (int i = 0; i < 5; i++)
        for (int j = 0; j < 5; j++) Inv[i][j] = a[i][5 + j];
    return 1;
}

static int qcoord(const int Inv[5][5], const int x[DIM]) {
    int y[5] = {0};
    for (int i = 0; i < 5; i++)
        for (int j = 0; j < 5; j++) y[i] = (y[i] + Inv[i][j] * x[j]) % 7;
    return y[2] * 49 + y[3] * 7 + y[4];
}

static int greedy_mis(int n, const uint64_t *adj, unsigned seed) {
    /* n=343, adj[i] is not a bitset of 343; use uint64_t adj[(n*n+63)/64] */
    int order[QN];
    for (int i = 0; i < n; i++) order[i] = i;
    unsigned rng = seed;
    for (int i = n - 1; i > 0; i--) {
        rng = rng * 1664525u + 1013904223u;
        int j = rng % (unsigned)(i + 1);
        int tmp = order[i];
        order[i] = order[j];
        order[j] = tmp;
    }
    uint8_t banned[QN];
    memset(banned, 0, n);
    int nt = 0;
    for (int t = 0; t < n; t++) {
        int v = order[t];
        if (banned[v]) continue;
        nt++;
        for (int u = 0; u < n; u++) {
            int bit = v * n + u;
            if (adj[bit / 64] & (1ull << (bit % 64))) banned[u] = 1;
        }
        banned[v] = 1;
    }
    return nt;
}

int main(void) {
    nsmall = 0;
    int o[5];
    for (o[0] = -1; o[0] <= 1; o[0]++)
    for (o[1] = -1; o[1] <= 1; o[1]++)
    for (o[2] = -1; o[2] <= 1; o[2]++)
    for (o[3] = -1; o[3] <= 1; o[3]++)
    for (o[4] = -1; o[4] <= 1; o[4]++) {
        for (int i = 0; i < 5; i++) SMALL[nsmall][i] = (o[i] + 7) % 7;
        nsmall++;
    }

    int best = 0, n_good = 0, checked = 0;
    uint64_t adj[(QN * QN + 63) / 64];

    for (int p0 = 0; p0 < 5; p0++)
    for (int p1 = p0 + 1; p1 < 5; p1++) {
        int free_idx[3], nf = 0;
        for (int j = 0; j < 5; j++)
            if (j != p0 && j != p1) free_idx[nf++] = j;
        int fill[6];
        int nfill = 1;
        for (int k = 0; k < 6; k++) nfill *= 7; /* 7^6 */
        for (int code = 0; code < nfill; code++) {
            int x = code;
            for (int k = 0; k < 6; k++) {
                fill[k] = x % 7;
                x /= 7;
            }
            int a[DIM] = {0}, b[DIM] = {0};
            a[p0] = 1;
            b[p1] = 1;
            a[free_idx[0]] = fill[0];
            a[free_idx[1]] = fill[1];
            a[free_idx[2]] = fill[2];
            b[free_idx[0]] = fill[3];
            b[free_idx[1]] = fill[4];
            b[free_idx[2]] = fill[5];
            checked++;
            if (!rank2_ok(a, b)) continue;
            if (!is_good(a, b)) continue;
            n_good++;
            int M[5][5], Inv[5][5];
            memset(M, 0, sizeof M);
            for (int j = 0; j < 5; j++) {
                M[0][j] = a[j];
                M[1][j] = b[j];
            }
            int ok = 0;
            for (int t0 = 0; t0 < 5 && !ok; t0++)
            for (int t1 = t0 + 1; t1 < 5 && !ok; t1++)
            for (int t2 = t1 + 1; t2 < 5 && !ok; t2++) {
                memset(M, 0, sizeof M);
                for (int j = 0; j < 5; j++) {
                    M[0][j] = a[j];
                    M[1][j] = b[j];
                }
                M[2][t0] = 1;
                M[3][t1] = 1;
                M[4][t2] = 1;
                if (invert5(M, Inv)) ok = 1;
            }
            if (!ok) continue;

            int forbidden[QN];
            memset(forbidden, 0, sizeof forbidden);
            int nforb = 0;
            int forblist[243];
            for (int s = 0; s < nsmall; s++) {
                int q = qcoord(Inv, SMALL[s]);
                if (!forbidden[q]) {
                    forbidden[q] = 1;
                    forblist[nforb++] = q;
                }
            }
            memset(adj, 0, sizeof adj);
            for (int i = 0; i < QN; i++) {
                int a0 = i / 49, r = i % 49, a1 = r / 7, a2 = r % 7;
                for (int fi = 0; fi < nforb; fi++) {
                    int f = forblist[fi];
                    if (f == 0) continue;
                    int b0 = f / 49, rr = f % 49, b1 = rr / 7, b2 = rr % 7;
                    int d0 = (a0 - b0 + 7) % 7;
                    int d1 = (a1 - b1 + 7) % 7;
                    int d2 = (a2 - b2 + 7) % 7;
                    int j = d0 * 49 + d1 * 7 + d2;
                    if (j == i) continue;
                    int bit = i * QN + j;
                    adj[bit / 64] |= 1ull << (bit % 64);
                }
            }
            int local_best = 0;
            for (int trial = 0; trial < 8; trial++) {
                int pack = greedy_mis(QN, adj, 1u + (unsigned)n_good * 17u + (unsigned)trial);
                if (pack > local_best) local_best = pack;
            }
            if (local_best > best) {
                best = local_best;
                printf("good #%d V=(%d%d%d%d%d / %d%d%d%d%d) forb=%d cosets=%d total=%d\n",
                       n_good, a[0], a[1], a[2], a[3], a[4], b[0], b[1], b[2], b[3],
                       b[4], nforb, local_best, 49 * local_best);
                fflush(stdout);
            }
            if (n_good <= 3 || n_good % 200 == 0) {
                printf("  good=%d checked=%d best_cosets=%d\n", n_good, checked, best);
                fflush(stdout);
            }
            if (best >= 8) {
                printf("HIT cosets=%d total=%d\n", best, 49 * best);
                /* materialising the set is left to Python if this prints */
            }
        }
    }
    printf("DONE checked=%d good=%d best_cosets=%d best_total=%d\n", checked, n_good,
           best, 49 * best);
    return 0;
}
